### DISCUSSION TITLE
Different semantic highlighting for builtin functions

### DISCUSSION THREAD

**Discussion by arrufat** [zigtools/zls#2104] [Ideas]
As I understand it, built-in functions are all highlighted the same way.

This makes sense for languages like Python, to help distinguish them from regular functions.

However, Zig is very explicit, due to the `@` prefix.

In my setup, I highlight them as regular functions, but it would be nice if built-in functions that return types could be highlighted as types. Some examples:

- `@intFromFloat`, `@typeInfo`: regular function
- `@Vector`, `@Type`: type functions, should be highlighted the same way as `std.ArrayList`, for example.

What do you think? Is this even feasible with the current design?

**Comment by nektro** [comment]
i think this is partially related to that the return value isnt resolved from any of the builtins atm. thus u dont get autocomplete either on the result

**Reply by techatrix** [reply]
> i think this is partially related to that the return value isnt resolved from any of the builtins atm.

This is inaccurate, ZLS can resolve the return value of various bultins.

**Comment by techatrix** [comment]
I think that the idea of special highlighting for different "kinds" of builtin functions is something that should be explored further at some point. There are some prerequisites that need to be resolved until we can get there. See https://github.com/zigtools/zls/issues/2106
